taichi/codegen/codegen_llvm.cpp

```diff
@@ -1313,37 +1314,80 @@ std::tuple<llvm::Value *, llvm::Value *> CodeGenLLVM::get_range_for_bounds(
 }
 
 void CodeGenLLVM::create_offload_struct_for(OffloadedStmt *stmt, bool spmd) {
-  llvm::Function *body;
+  // TODO: instead of constructing tons of LLVM IR, writing the logic in
+  // runtime.cpp may be a cleaner solution. See
+  // CodeGenLLVMCPU::create_offload_range_for as an example.
+
+  llvm::Function *body = nullptr;
   auto leaf_block = stmt->snode;
   {
     // Create the loop body function
     auto guard = get_function_creation_guard({
         llvm::PointerType::get(get_runtime_type("Context"), 0),
+        get_tls_buffer_type(),
         llvm::PointerType::get(get_runtime_type("Element"), 0),
         tlctx->get_data_type<int>(),
         tlctx->get_data_type<int>(),
     });
 
     body = guard.body;
 
-    // per-leaf-block for loop
+    /* Function structure:
+     *
+     * function_body (entry):
+     *   loop_index = lower_bound;
+     *   bls_prologue()
+     *   goto loop_test
+     *
+     * loop_test:
+     *   if (loop_index < upper_bound)
+     *     goto loop_body
+     *   else
+     *     goto func_exit
+     *
+     * loop_body:
+     *   initialize_coordinates()
+     *   if (bitmasked voxel is active)
+     *     goto struct_for_body
+     *   else
+     *     goto loop body tail
+     *
+     * struct_for_body:
+     *   ... (Run codegen on the StructForStmt::body Taichi Block)
+     *   goto loop_body_tail
+     *
+     * loop_body_tail:
+     *   loop_index += block_dim
+     *   goto loop_test
+     *
+     * func_exit:
+     *   bls_epilogue()
+     *   return
+     */
+
     auto loop_index =
         create_entry_block_alloca(llvm::Type::getInt32Ty(*llvm_context));
 
-    llvm::Value *thread_idx = nullptr, *block_dim = nullptr;
+    RuntimeObject element("Element", this, builder.get(), get_arg(2));
 
-    RuntimeObject element("Element", this, builder.get(), get_arg(1));
-    auto lower_bound = get_arg(2);
-    auto upper_bound = get_arg(3);
+    // Loop ranges
+    auto lower_bound = get_arg(3);
+    auto upper_bound = get_arg(4);
```

Comment:
Now we use `arg(2)` for the TLS buffer base pointer, hence the argument id offset for lower/upper bound.